libcontainer/container_linux.go

```diff
@@ -267,20 +268,64 @@ func (c *linuxContainer) Exec() error {
 
 func (c *linuxContainer) exec() error {
 	path := filepath.Join(c.root, execFifoFilename)
-	f, err := os.OpenFile(path, os.O_RDONLY, 0)
-	if err != nil {
-		return newSystemErrorWithCause(err, "open exec fifo for reading")
+
+	select {
+	case <-awaitProcessExit(c.initProcess.pid()):
+		return errors.New("container process is already dead")
+	case result := <-awaitFifoOpen(path):
+		if result.err != nil {
+			return result.err
+		}
+		f := result.file
+		defer f.Close()
+		if err := readFromExecFifo(f); err != nil {
+			return err
+		}
+		return os.Remove(path)
```

Comment:
should the file be closed first or does it not matter?

Reply:
Not sure why this should matter. I suppose the fd will point to an inaccessible location on the filesystem for some amount of time, but in terms of code cleanliness, the `defer` seems better?